### DISCUSSION TITLE
Need help to implement "+" animation for nested nodes

### DISCUSSION THREAD

**Discussion by sathya-1994** [eclipse-glsp/glsp#452] [General] [answered]
I am using GLSP's vscode extension(https://github.com/eclipse-glsp/glsp-vscode-integration) and language server which extends GLSP server.

I am creating nodes using GModel. 
**Screenshot:**
![image](https://user-images.githubusercontent.com/59145404/140680883-55aebc98-219f-42c8-b901-b569d013007c.png)

I want to add + button to display/hide nested nodes. How we can implement this feature?
Or Already this implementation is available in GLSP?
**Screenshot:**
![image](https://user-images.githubusercontent.com/59145404/140681350-bb73c0f5-f175-4751-95f1-254fb929be4e.png)

**My idea:** When user click "+" icon, we will call layout operation and we will add/remove child/nested nodes while layout generation based on the boolean value(new filed added to model state) stored in the model state(org.eclipse.glsp.server.model.GModelState). We will update this new filed when user clicks "+" icon.

**Graph hierarchy:**
![image](https://user-images.githubusercontent.com/59145404/140683156-eed881ff-efce-4df0-9822-acac1a4c68d1.png)

**Comment by planger** [comment, marked as answer]
There are a few different approaches to implement such a button feature:
- Add a placeholder node (GModelElement, such as a label etc) to the GModel and register a client-side click listener, which sends whatever action you need on click
- Register a client-side UIExtension (HTML) that is rendered on top of the diagram on each element, which will again on click send whatever action you need
- Use GButton, which is a GModelElement. I haven't used this for quite some time though.
